### PR TITLE
Add getters on BoundStatement (JAVA-547)

### PR DESCRIPTION
**Pull request by olim7t**
Backport `BoundStatement.get<Type>()` from 2.1, and add a global `getValues()`.

This is depended upon by JAVA-646 -- slow query logger (not for the feature itself, but for `DataType#format` that I also backported from 2.1).

### CODE REVIEW

File: driver-core/src/main/java/com/datastax/driver/core/BoundStatement.java [hunk 1]
```diff
@@ -1104,6 +1117,199 @@ public BoundStatement setPartitionKeyToken(Token v) {
...
+     * Returns all the values bound for this statement.
+     * <p>
+     * Note that this method may have a non-negligible impact on performance: internally, values
+     * are stored in serialized form, so they need to be deserialized one by one. This method is
+     * provided for debugging purposes.
+     *
+     * @return the values, in the order in which they appear in the statement. Any unset value
+     * will be returned as {@code null}.
+     */
+    public List<Object> getValues() {
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/BoundStatement.java]: I wonder: wouldn't it be even more useful to return a LinkedHashMap<String,Object> where keys would be the parameter name and values the corresponding value? Implementation-wise the name is retrievable via metadata().getName(i); I'm suggesting LinkedHashMap to preserve the order in which parameters appear in the statement.

Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/BoundStatement.java]: Also: would it make sense to declare this method at GettableData interface level? An implementation in AbstractGettableData seems feasible and would benefit to the Row interface too.

Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/BoundStatement.java]: Yes, both very good points. I need to check how names work when the query uses positional parameters, but IIRC they get generated.

Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/BoundStatement.java]: From my experience positional parameters use original column names as parameter names - but indeed this detail needs to be double-checked.

Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/BoundStatement.java]: True, but a column can also appear twice, which is where this is going to get complicated.

**Comment by olim7t** [comment]
@adutra I've moved getValues to the GettableData interface. It's still a list because of the potential duplicate identifiers, if clients need more details they can downcast to Row or BoundStatement to get the ColumnDefinitions.
I also figured it would be a good time to add getObject since it's closely related.

**Comment by adutra** [comment]
@olim7t Awesome!

**Pull request merged**